### ISSUE TITLE
Use of String.to_atom/1 opens up DoS possibility

### ISSUE THREAD

**Issue by Nicd** [cnavas88/exvalidate#21]
The project uses String.to_atom/1 in these places:

https://github.com/cnavas88/exvalidate/blob/master/lib/exvalidate.ex#L117

https://github.com/cnavas88/exvalidate/blob/master/lib/workflow/rules/type.ex#L208

Using `String.to_atom/1` on dynamic or user-provided input opens the possibility for denial of service attacks. The issue stems from the fact that the BEAM atom table has a fixed maximum size. If this size is exceeded, the whole VM crashes. When a user provides strings that are converted to atoms, they an easily overflow the atom table.

The alternative is to use `String.to_existing_atom/1`, which will only support atoms that already exist in the atom table (i.e. have been already used in the code). Or the code needs to be structured to use e.g. strings instead of atoms.

**Comment by cnavas88** [comment]
As you say a solution would be to use strings instead of atoms. How about the solution of rescuing the error of giving the BEAM over the limit of the atoms?

`x in SystemLimitError -> ....`

**Comment by Nicd** [comment]
I don't think you can catch that error. And even if you can, you've still filled the atom table. The table is not garbage collected so your application would just die in another place.

I think since the user defines the schema beforehand and that includes the atoms, `String.to_existing_atom/1` would serve you well here.

**Comment by cnavas88** [comment]
perfect, thanks for your help!! 

I will think about all the possibilities you mention and I will make tests to see which is the best and if there can be another possibility for this case.

feel free to change what you think and to keep making the mistakes you see.
